diff --git a/dlls/riched20/writer.c b/dlls/riched20/writer.c
index 70d5290b01..58449d1144 100644
--- a/dlls/riched20/writer.c
+++ b/dlls/riched20/writer.c
@@ -686,8 +686,8 @@ static BOOL stream_out_para_props( ME_TextEditor *editor, ME_OutStream *pStream,
             /* Word bar tab (vertical bar). Handled below */
             break;
         }
-        if (fmt->rgxTabs[i] >> 28 <= 5)
-          strcat(props, leader[fmt->rgxTabs[i] >> 28]);
+        if ((ULONG)fmt->rgxTabs[i] >> 28 <= 5)
+          strcat(props, leader[(ULONG)fmt->rgxTabs[i] >> 28]);
         sprintf(props+strlen(props), "\\tx%ld", fmt->rgxTabs[i]&0x00FFFFFF);
       }
     }
