diff --git a/dlls/ws2_32/socket.c b/dlls/ws2_32/socket.c
index 95cb305902..f5580cab4e 100644
--- a/dlls/ws2_32/socket.c
+++ b/dlls/ws2_32/socket.c
@@ -1353,7 +1353,7 @@ int WINAPI closesocket( SOCKET s )
         return -1;
     }
 
-    if (!socket_list_remove( s ))
+    if (!socket_list_remove( s ) && !is_valid_socket( s ))
     {
         SetLastError( WSAENOTSOCK );
         return -1;
diff --git a/dlls/ws2_32/tests/sock.c b/dlls/ws2_32/tests/sock.c
index 8d589c63ea..d861f30336 100644
--- a/dlls/ws2_32/tests/sock.c
+++ b/dlls/ws2_32/tests/sock.c
@@ -14595,10 +14595,16 @@ static void test_valid_handle(void)
     ok(ret == -1, "got %d\n", ret);
     ok(WSAGetLastError() == WSAENOTSOCK, "got error %u\n", WSAGetLastError());
 
-    CloseHandle(invalid);
-    CloseHandle(duplicated);
-    closesocket(client);
-    closesocket(server);
+    ret = closesocket((SOCKET)invalid);
+    ok(ret == SOCKET_ERROR, "got %d, expected SOCKET_ERROR.\n", ret);
+    ret = CloseHandle(invalid);
+    ok(ret, "CloseHandle failed unexpectedly: %d\n", ret);
+    ret = closesocket((SOCKET)duplicated);
+    ok(!ret, "closesocket failed unexpectedly: %d\n", ret);
+    ret = closesocket(client);
+    ok(!ret, "closesocket failed unexpectedly: %d\n", ret);
+    ret = closesocket(server);
+    ok(!ret, "closesocket failed unexpectedly: %d\n", ret);
 }
 
 static void test_afunix_path( const char *path )
