diff --git a/dlls/advapi32/eventlog.c b/dlls/advapi32/eventlog.c
index 46300bd9e6..549e0e97e8 100644
--- a/dlls/advapi32/eventlog.c
+++ b/dlls/advapi32/eventlog.c
@@ -618,7 +618,10 @@ BOOL WINAPI ReportEventA ( HANDLE hEventLog, WORD wType, WORD wCategory, DWORD d
     ret = ReportEventW(hEventLog, wType, wCategory, dwEventID, lpUserSid,
                        wNumStrings, dwDataSize, (LPCWSTR *)wideStrArray, lpRawData);
     for (i = 0; i < wNumStrings; i++)
-        free(wideStrArray[i]);
+    {
+        str.Buffer = wideStrArray[i];
+        RtlFreeUnicodeString(&str);
+    }
     free(wideStrArray);
     return ret;
 }
