diff --git a/dlls/msvcp90/exception.c b/dlls/msvcp90/exception.c
index 01a366078b..1922a179ba 100644
--- a/dlls/msvcp90/exception.c
+++ b/dlls/msvcp90/exception.c
@@ -1302,10 +1302,7 @@ void __cdecl __ExceptionPtrAssign(exception_ptr *ep, const exception_ptr *assign
 {
     TRACE("(%p %p)\n", ep, assign);
 
-    /* don't destroy object stored in ep */
-    if (ep->ref)
-        InterlockedDecrement(ep->ref);
-
+    __ExceptionPtrDestroy(ep);
     *ep = *assign;
     if (ep->ref)
         InterlockedIncrement(ep->ref);
diff --git a/dlls/msvcrt/exception_ptr.c b/dlls/msvcrt/exception_ptr.c
index 383d16166a..4524f718c1 100644
--- a/dlls/msvcrt/exception_ptr.c
+++ b/dlls/msvcrt/exception_ptr.c
@@ -116,10 +116,7 @@ void __cdecl __ExceptionPtrAssign(exception_ptr *ep, const exception_ptr *assign
 {
     TRACE("(%p %p)\n", ep, assign);
 
-    /* don't destroy object stored in ep */
-    if (ep->ref)
-        InterlockedDecrement(ep->ref);
-
+    __ExceptionPtrDestroy(ep);
     *ep = *assign;
     if (ep->ref)
         InterlockedIncrement(ep->ref);
