diff --git a/dlls/nsiproxy.sys/ip.c b/dlls/nsiproxy.sys/ip.c
index 2b724db8da..6c4d5808be 100644
--- a/dlls/nsiproxy.sys/ip.c
+++ b/dlls/nsiproxy.sys/ip.c
@@ -1228,6 +1228,12 @@ static NTSTATUS ipv4_neighbour_enumerate_all( void *key_data, UINT key_size, voi
                                      NULL, 0, NULL, 0, &iface_count )))
         return status;
 
+    if (!iface_count)
+    {
+        *count = 0;
+        return STATUS_SUCCESS;
+    }
+
     if (!(luid_tbl = malloc( iface_count * sizeof(*luid_tbl) )))
         return STATUS_NO_MEMORY;
 
