diff --git a/dlls/mshtml/htmlelem.c b/dlls/mshtml/htmlelem.c
index 4db8bd83a3..ab643a74d8 100644
--- a/dlls/mshtml/htmlelem.c
+++ b/dlls/mshtml/htmlelem.c
@@ -8209,6 +8209,7 @@ static HRESULT HTMLAttributeCollection_get_dispid(DispatchEx *dispex, const WCHA
             if(iter == nsattr)
                 break;
         }
+        nsIDOMAttr_Release(nsattr);
 
         assert(i < length);
         *dispid = MSHTML_DISPID_CUSTOM_MIN + i;
