diff --git a/dlls/crypt32/oid.c b/dlls/crypt32/oid.c
index a35917ff8f..7c6c07e637 100644
--- a/dlls/crypt32/oid.c
+++ b/dlls/crypt32/oid.c
@@ -740,15 +740,31 @@ BOOL WINAPI CryptRegisterOIDInfo(PCCRYPT_OID_INFO info, DWORD flags)
     char *key_name;
     HKEY root = 0, key = 0;
     DWORD err;
+    CRYPT_OID_INFO info_copy = {0};
 
     TRACE("(%p, %lx)\n", info, flags );
 
-    if (!info || info->cbSize != sizeof(*info) || !info->pszOID)
+    if (!info || !info->pszOID)
     {
         SetLastError(E_INVALIDARG);
         return FALSE;
     }
 
+    if (info->cbSize != sizeof(*info)) {
+#ifdef CRYPT_OID_INFO_HAS_EXTRA_FIELDS
+        if (info->cbSize == sizeof(CRYPT_OID_INFO) - sizeof(LPWSTR) * 2) {
+            memcpy(&info_copy, info, sizeof(CRYPT_OID_INFO) - sizeof(LPWSTR) * 2);
+            info_copy.cbSize = sizeof(CRYPT_OID_INFO);
+            info = &info_copy;
+        }
+        else
+#endif
+        {
+            SetLastError(E_INVALIDARG);
+            return FALSE;
+        }
+    }
+
     if (!info->dwGroupId) return TRUE;
 
     key_name = CryptMemAlloc(strlen(info->pszOID) + 16);
diff --git a/dlls/crypt32/tests/oid.c b/dlls/crypt32/tests/oid.c
index 7206668be1..9a3376caf1 100644
--- a/dlls/crypt32/tests/oid.c
+++ b/dlls/crypt32/tests/oid.c
@@ -36,6 +36,19 @@ struct OIDToAlgID
     DWORD altAlgID;
 };
 
+struct CRYPT_OID_INFO_XP {
+    DWORD   cbSize;
+    LPCSTR  pszOID;
+    LPCWSTR pwszName;
+    DWORD   dwGroupId;
+    union {
+        DWORD  dwValue;
+        ALG_ID Algid;
+        DWORD  dwLength;
+    } DUMMYUNIONNAME;
+    CRYPT_DATA_BLOB ExtraInfo;
+};
+
 static const struct OIDToAlgID oidToAlgID[] = {
  { szOID_RSA_RSA, CALG_RSA_KEYX },
  { szOID_RSA_MD2RSA, CALG_MD2 },
@@ -592,6 +605,7 @@ static void test_registerOIDInfo(void)
 {
     static char test_oid[] = "1.2.3.4.5.6.7.8.9.10";
     CRYPT_OID_INFO info1;
+    struct CRYPT_OID_INFO_XP info_xp = {0};
     const CRYPT_OID_INFO *info2;
     HKEY key;
     DWORD ret, size, type, value;
@@ -631,6 +645,12 @@ static void test_registerOIDInfo(void)
     ret = CryptRegisterOIDInfo(&info1, 0);
     ok(ret, "got %lu\n", GetLastError());
 
+    memcpy(&info_xp, &info1, sizeof(info_xp));
+    info_xp.cbSize = sizeof(info_xp);
+
+    ret = CryptRegisterOIDInfo((CRYPT_OID_INFO*)&info_xp, 0);
+    ok(ret, "got %lu\n", GetLastError());
+
     ret = RegOpenKeyA(HKEY_LOCAL_MACHINE, "Software\\Microsoft\\Cryptography\\OID\\EncodingType 0\\CryptDllFindOIDInfo\\1.2.3.4.5.6.7.8.9.10!1", &key);
     ok(ret == ERROR_FILE_NOT_FOUND, "got %lu\n", ret);
 
